geowebcache/sqlite/src/main/java/org/geowebcache/sqlite/FileManager.java

```diff
@@ -0,0 +1,376 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * <p>
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Nuno Oliveira, GeoSolutions S.A.S., Copyright 2016
+ */
+package org.geowebcache.sqlite;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.geowebcache.storage.TileObject;
+import org.geowebcache.storage.TileRange;
+import org.geowebcache.storage.blobstore.file.FilePathGenerator;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.geowebcache.sqlite.Utils.Tuple;
+
+/**
+ * Class responsible to map GWC concepts (layer, tile, tile range, etc  ...)
+ * to a filesystem file. The mapping is defined by a template that can use
+ * the information associated with a tile.
+ * <p>
+ * <p/>
+ * The template supported terms are:
+ * <ul>
+ * <li>params</li>
+ * <li>x</li>
+ * <li>y</li>
+ * <li>z</li>
+ * <li>layer</li>
+ * <li>grid</li>
+ * <li>format</li>
+ * </ul>
+ * It is also possible to use parameters referencing them by their name.
+ * </p>
+ * <p>
+ * <p>
+ * For example a template like the following:
+ * <blockquote><pre>
+ * {grid}/{layer}/{format}/{params}/{z}/tiles_{x}_{y}.sqlite
+ * </pre></blockquote>
+ * will produce paths similar to this one:
+ * <blockquote><pre>
+ * EPSG_4326/img_states/image_png/10/tiles_350_625.sqlite
+ * </pre></blockquote>
+ * </p>
+ * <p>
+ * <p>
+ * Is possible to map all tiles to a single file by defining a static template (no terms).
+ * Although, if a term is used it cannot be NULL otherwise an exception will be throw.
+ * If a referenced parameter doesn't exists the string 'null' will be used.
+ * </p>
+ */
+final class FileManager {
+
+    private static Log LOGGER = LogFactory.getLog(FileManager.class);
+
+    private final static Pattern PATH_TEMPLATE_ATTRIBUTE_PATTERN = Pattern.compile("\\{(.+?)\\}");
+
+    private final File rootPath;
+
+    private final long rowRangeCount;
+    private final long columnRangeCount;
+
+    // path builder extracted from the path template
+    private final String[] pathBuilderOriginal;
+    // thread local copy of the path builder to avoid instantiate a new one each time
+    private final ThreadLocal<String[]> pathBuilderLocal = new ThreadLocal<>();
```

Comment:
Doh, I was wrong though, thread locals should be cleaned at each request end, otherwise they will be left attached to the threads and prevent clean undeploy.